op-batcher/batcher/driver.go

```diff
@@ -482,6 +489,32 @@ func (l *BatchSubmitter) safeL1Origin(ctx context.Context) (eth.BlockID, error)
 	return status.SafeL2.L1Origin, nil
 }
 
+// fallbackTxCandidate creates a tx candidate for the given txdata using either
+// 4844 blobs or calldata based on gas fee. It throws an error if fallback is
+// disabled.
+func (l *BatchSubmitter) fallbackTxCandidate(ctx context.Context, txdata txData) (*txmgr.TxCandidate, error) {
+	if l.DAClient.EthFallbackDisabled {
+		return nil, fmt.Errorf("celestia: fallback disabled")
+	}
+	_, baseFee, blobBaseFee, err := l.Txmgr.SuggestGasPriceCaps(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not get gas price info: %w", err)
+	}
+	blobBaseFee.Mul(blobBaseFee, blobTxBlobGasPerBlob)
+	blobBaseFee.Div(blobBaseFee, usableBytesInBlob)
+	calldataFeePerByte := baseFee.Mul(baseFee, big.NewInt(16))
+	if blobBaseFee.Cmp(calldataFeePerByte) <= 0 {
```

Comment:
having a switch based on fee is interesting and probably a feature for upstream optimism

for simplicity i'd fallback only on blob or calldata, not both